test/jqLiteSpec.js

```diff
@@ -594,6 +594,39 @@ describe('jqLite', function() {
         }).not.toThrow();
       });
     });
+
+    fdescribe('camelCasing keys', function() {
+      // jQuery 2.x has different behavior; skip the tests.
+      // if (isJQuery2x()) return;
+
+      it('should camelCase the key in a setter', function() {
+        var element = jqLite(a);
+
+        element.data('a-B-c-d-42--e', 'z-x');
+        expect(element.data()).toEqual({'a-BCD-42-E': 'z-x'});
```

Comment:
This is also how the HTML5 spec converts data attribute names to keys in dataset. That's where jQuery took it from.

It's also nice in that it's fully 1-1, the conversion doesn't lose any information.